adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -193,31 +211,42 @@ public final void stop(Future<Void> stopFuture) throws Exception {
             }
         });
 
-        Future<Void> serverTracker = Future.future();
+        Future<Void> serverStopTracker = Future.future();
         if (server != null) {
-            server.close(serverTracker.completer());
+            server.close(serverStopTracker.completer());
         } else {
-            serverTracker.complete();
+            serverStopTracker.complete();
         }
-        Future<Void> afterServerStopTracker = Future.future();
-        serverTracker.compose(d -> {
-            afterHttpServerStop(afterServerStopTracker);
-        }, afterServerStopTracker);
 
-        afterServerStopTracker.compose(d -> {
+        Future<Void> honoClientStopTracker = Future.future();
+        serverStopTracker.compose(d -> {
             if (hono != null) {
-                hono.shutdown(shutdownTracker.completer());
+                hono.shutdown(honoClientStopTracker.completer());
             } else {
-                shutdownTracker.complete();
+                honoClientStopTracker.complete();
             }
+        }, honoClientStopTracker);
+
+        honoClientStopTracker.compose(d -> {
+            postShutdown(shutdownTracker);
         }, shutdownTracker);
     }
 
     /**
-     * Invoked after the Http server is stopped.
-     * May be overridden by sub-classes to provide custom stop handling.
+     * Invoked before the Http server is shut down.
+     * May be overridden by sub-classes.
+     */
+    protected void preShutdown() {
```

Comment:
Corresponding to `postStartup()`, I would catch and log exceptions there.
In any case, `postStartup` should not prevent shutdown of the httpServer and honoClient.